docs/examples/distributed/002_multi_gpu/main.py

```diff
@@ -0,0 +1,271 @@
+
+    # NOTE: Here `batch_size` is still the "local" (per-gpu) batch size.
+    # This way, the effective batch size scales directly with number of GPUs, no need to specify it
+    # in advance. You might want to adjust the learning rate and other hyper-parameters though.
+    if is_master:
+        logger.info(f"Effective batch size: {batch_size * world_size}")
+    train_dataloader = DataLoader(
+        train_dataset,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        shuffle=False,  # shuffling is now done in the sampler, not the dataloader.
+        sampler=train_sampler,
+    )
+    valid_dataloader = DataLoader(
+        valid_dataset,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        shuffle=False,
+        sampler=valid_sampler,
+    )
+    test_dataloader = DataLoader(  # NOTE: Not used in this example.
+        test_dataset,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        shuffle=False,
+        sampler=test_sampler,
+    )
+
+    # Checkout the "checkpointing and preemption" example for more info!
+    logger.debug("Starting training from scratch.")
+
+    for epoch in range(training_epochs):
+        logger.debug(f"Starting epoch {epoch}/{training_epochs}")
+
+        # NOTE: Here we need to call `set_epoch` so the ordering is able to change at each epoch.
+        train_sampler.set_epoch(epoch)
+
+        # Set the model in training mode (important for e.g. BatchNorm and Dropout layers)
+        model.train()
+
+        # NOTE: using a progress bar from tqdm because it's nicer than using `print`.
+        progress_bar = tqdm(
+            total=len(train_dataloader),
+            desc=f"Train epoch {epoch}",
+            disable=not is_master,
+        )
+
+        # Training loop
+        for batch in train_dataloader:
+            # Move the batch to the GPU before we pass it to the model
+            batch = tuple(item.to(device) for item in batch)
+            x, y = batch
+
+            # Forward pass
+            logits: Tensor = model(x)
+
+            local_loss = F.cross_entropy(logits, y)
+
+            optimizer.zero_grad()
+            local_loss.backward()
+            # NOTE: nn.DistributedDataParallel automatically averages the gradients across devices.
+            optimizer.step()
+
+            # Calculate some metrics:
+            # local metrics
+            local_n_correct_predictions = logits.detach().argmax(-1).eq(y).sum()
+            local_n_samples = logits.shape[0]
+            local_accuracy = local_n_correct_predictions / local_n_samples
+
+            # "global" metrics: calculated with the results from all workers
+            # NOTE: Creating new tensors to hold the "global" values, but this isn't required.
+            n_correct_predictions = local_n_correct_predictions.clone()
+            torch.distributed.all_reduce(n_correct_predictions, op=torch.distributed.ReduceOp.SUM)
```

Comment:
Probably not that impactful but
Because only the master needs the final value for logging purposes you could switch `all_reduce` to `reduce`
that might avoid some unnecessary data transfer

Reply:
Fixed in 7c0a5f5b76ab65c49f7333f8222d51018d89e7a8

Reply:
Actually fixed in fc0a8d910520dd9a911c1d6c94539b080200c138